### PR TITLE
Improve namer.GetName

### PR DESCRIPTION
**Pull request by rhcarvalho**
* Return early if `len(name) <= maxLength`
* Avoid "slice bounds out of range" panic if used with small `maxLength` values
* Guarantee that the length of the return value is always less than or equal `maxLength`

### CODE REVIEW

**Comment by rhcarvalho** [comment]
[test]

**Comment by rhcarvalho** [comment]
@deads2k PTAL

**Comment by rhcarvalho** [comment]
jenkins failure is a flake, #6447

File: pkg/util/namer/namer.go [hunk 1]
```diff
@@ -13,23 +13,27 @@ import (
 // an 8-character hash of the [base]-[suffix] string.  If the suffix is not too long,
 // it will truncate the base, add the hash of the base and return [base]-[hash]-[suffix]
 func GetName(base, suffix string, maxLength int) string {
+	if maxLength <= 0 {
```
Comment (mfojtik) [inline pkg/util/namer/namer.go]: why is this needed?

File: pkg/util/namer/namer_test.go [hunk 2]
```diff
@@ -77,6 +76,19 @@ func TestGetNameIsDifferent(t *testing.T) {
...
 }
 
+func TestGetNameReturnShortNames(t *testing.T) {
+	base := randSeq(32)
+	for maxLength := 0; maxLength < len(base)+2; maxLength++ {
+		for suffixLen := 0; suffixLen <= maxLength+1; suffixLen++ {
+			suffix := randSeq(suffixLen)
+			got := GetName(base, suffix, maxLength)
+			if len(got) > maxLength {
+				t.Fatalf("len(GetName(%[1]q, %[2]q, %[3]d)) = len(%[4]q) = %[5]d; want %[3]d", base, suffix, maxLength, got, len(got))
```
Comment (deads2k) [inline pkg/util/namer/namer_test.go]: For looping tests, use `t.Errorf` instead so additional tests run.

**Comment by deads2k** [comment]
Minor comments, squash commits, and this can merge post-1.1.1

File: pkg/util/namer/namer.go [hunk 3]
```diff
@@ -13,23 +13,27 @@ import (
...
-		}
+	baseLength := maxLength - 10 /*length of -hash-*/ - len(suffix)
 
-		prefix := base[0:baseLength]
+	// if the suffix is too long, ignore it
+	if baseLength < 0 {
+		prefix := base[0:min(len(base), max(0, maxLength-9))]
 		// Calculate hash on initial base-suffix string
-		return fmt.Sprintf("%s-%s-%s", prefix, hash(base), suffix)
+		shortName := fmt.Sprintf("%s-%s", prefix, hash(name))
```
Comment (mfojtik) [inline pkg/util/namer/namer.go]: since we use this on many places, thus we care about performance and you are already here... can you replace this with `strings.Join` ?

File: pkg/util/namer/namer.go [hunk 4]
```diff
@@ -13,23 +13,27 @@ import (
...
+		return shortName[:min(maxLength, len(shortName))]
 	}
 
-	return name
+	prefix := base[0:baseLength]
+	// Calculate hash on initial base-suffix string
+	return fmt.Sprintf("%s-%s-%s", prefix, hash(base), suffix)
```
Comment (mfojtik) [inline pkg/util/namer/namer.go]: `strings.Join()` here pls.

File: pkg/util/namer/namer.go (hunk 3 above)
Comment (deads2k) [inline pkg/util/namer/namer.go]: > since we use this on many places, thus we care about performance and you are already here... can you replace this with strings.Join ?

Has this surfaced on a profile?  I find `strings.Join` hard to read and this method is called right before a remote http call, so I wouldn't expect it to be high in the list.

Comment (mfojtik) [inline pkg/util/namer/namer.go]: @deads2k it is not on profile, but it is definitely less expensive than string formatting :-)

File: pkg/util/namer/namer_test.go (hunk 2 above)
Comment (rhcarvalho) [inline pkg/util/namer/namer_test.go]: I used `t.Fatalf` here on purpose, as there's no benefit to see hundreds of lines with random strings in the output.

File: pkg/util/namer/namer.go (hunk 1 above)
Comment (rhcarvalho) [inline pkg/util/namer/namer.go]: precondition checking

**Comment by rhcarvalho** [comment]
@deads2k @mfojtik commits squashed, PTAL

**Comment by rhcarvalho** [comment]
Jenkins is shouting "Warning: you have no plugins providing access control for builds, so falling back to legacy behavior of permitting any downstream builds to be triggered
Finished: FAILURE", and apparently no tests were run.

[test]

**Comment by rhcarvalho** [comment]
Jenkins flake https://github.com/openshift/origin/issues/6437

File: pkg/util/namer/namer.go (hunk 1 above)
Comment (rhcarvalho) [inline pkg/util/namer/namer.go]: If we don't do this check, code later will compute lengths and slice indices incorrectly for certain inputs.

**Comment by rhcarvalho** [comment]
[test]

**Comment by rhcarvalho** [comment]
@bparees PTAL

**Comment by bparees** [comment]
lgtm [merge]
